pkg/scheduler/plugins/elasticquota/core/quota_tree_wrapper.go

```diff
@@ -0,0 +1,481 @@
+type quotaTreeMapType map[string]*quotaTree
+
+type QuotaTreeWrapper struct {
+	globalRuntimeVersion int64
+	resourceKeys         map[string]struct{}
+	groupReqLimit        quotaResMapType
+	quotaTree            quotaTreeMapType
+	totalResource        v1.ResourceList
+	lock                 sync.RWMutex
+	treeName             string
+}
+
+func NewQuotaTreeWrapper(treeName string) *QuotaTreeWrapper {
+	return &QuotaTreeWrapper{
+		globalRuntimeVersion: 1,
+		resourceKeys:         make(map[string]struct{}),
+		groupReqLimit:        make(quotaResMapType),
+		quotaTree:            make(quotaTreeMapType),
+		totalResource:        v1.ResourceList{},
+		treeName:             treeName,
+	}
+}
+
+func (qtw *QuotaTreeWrapper) LockUpdateResourceKeys(resourceKeys map[string]struct{}) {
+	newResourceKey := make(map[string]struct{})
+	for resKey := range resourceKeys {
+		newResourceKey[resKey] = struct{}{}
+	}
+
+	qtw.lock.Lock()
+	defer qtw.lock.Unlock()
+
+	qtw.resourceKeys = newResourceKey
+	qtw.updateQuotaTreeDimensionByResourceKeysNoLock()
+}
+
+func (qtw *QuotaTreeWrapper) updateQuotaTreeDimensionByResourceKeysNoLock() {
+	//lock outside
+	for resKey := range qtw.quotaTree {
+		if _, exist := qtw.resourceKeys[resKey]; !exist {
+			delete(qtw.quotaTree, resKey)
+		}
+	}
+
+	for resKey := range qtw.resourceKeys {
+		if _, exist := qtw.quotaTree[resKey]; !exist {
+			qtw.quotaTree[resKey] = NewQuotaTree()
+		}
+	}
+}
+
+func (qtw *QuotaTreeWrapper) LockDeleteOneGroup(quotaInfo *QuotaInfo) {
+	qtw.lock.Lock()
+	defer qtw.lock.Unlock()
+
+	if klog.V(5).Enabled() {
+		qtw.LogWithUpdateOneGroupQuotaInfoNoLock("DeleteOneGroup before", quotaInfo)
+		defer qtw.LogWithUpdateOneGroupQuotaInfoNoLock("DeleteOneGroup after", quotaInfo)
+	}
+
+	for key := range qtw.resourceKeys {
+		qtw.quotaTree[key].erase(quotaInfo.Name)
+	}
+
+	delete(qtw.groupReqLimit, quotaInfo.Name)
+	qtw.adjustQuotaNoLock()
+}
+
+func (qtw *QuotaTreeWrapper) LockUpdateOneGroupMaxQuota(quotaInfo *QuotaInfo, newMaxQuota v1.ResourceList) {
+	qtw.lock.Lock()
+	defer qtw.lock.Unlock()
+
+	if klog.V(5).Enabled() {
+		qtw.LogWithUpdateOneGroupQuotaInfoNoLock("UpdateOneGroupMaxQuota before", quotaInfo)
+		defer qtw.LogWithUpdateOneGroupQuotaInfoNoLock("UpdateOneGroupMaxQuota after", quotaInfo)
+	}
+
+	for resKey := range newMaxQuota {
+		key := string(resKey)
+		qtw.resourceKeys[key] = struct{}{}
+		if _, exist := qtw.quotaTree[key]; !exist {
+			qtw.quotaTree[key] = NewQuotaTree()
+		}
+	}
+
+	reqLimit := qtw.getGroupRequestLimitNoLock(quotaInfo.Name, true)
+	for key := range qtw.resourceKeys {
+		resKey := v1.ResourceName(key)
+		//update reqLimitPerKey
+		reqLimitPerKey := *quotaInfo.CalculateInfo.Request.Name(resKey, resource.DecimalSI)
+		newMaxQuotaPerKey := *newMaxQuota.Name(resKey, resource.DecimalSI)
+		if reqLimitPerKey.Cmp(newMaxQuotaPerKey) == 1 {
+			reqLimitPerKey = newMaxQuotaPerKey.DeepCopy()
+		}
+		reqLimit[resKey] = reqLimitPerKey
```

Comment:
![image](https://user-images.githubusercontent.com/70878006/185937400-b391d37f-8f57-4b0d-bd50-533d1027d2ed.png)

use “\n” to seperate different logic to impove reading effect.